### PR TITLE
guix: Build support for macOS

### PR DESCRIPTION
**Pull request by dongcarl**
This PR brings our Guix builds on par with Gitian in terms of supported architectures.

Reviewers: if you run a build, please submit:

```
find output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
```

So that we can compare hashes and ensure reproducibility!

### CODE REVIEW

File: contrib/guix/libexec/build-osx.sh [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+#         i686-linux-gnu)      echo /lib/ld-linux.so.2 ;;
+#         x86_64-linux-gnu)    echo /lib64/ld-linux-x86-64.so.2 ;;
+#         arm-linux-gnueabihf) echo /lib/ld-linux-armhf.so.3 ;;
+#         aarch64-linux-gnu)   echo /lib/ld-linux-aarch64.so.1 ;;
+#         riscv64-linux-gnu)   echo /lib/ld-linux-riscv64-lp64d.so.1 ;;
+#         *)                   exit 1 ;;
+#     esac
+# )
+
+# Environment variables for determinism
```
Comment (fanquake) [inline contrib/guix/libexec/build-osx.sh]: I assume we no longer have to set `ZERO_AR_DATE` here because we're not using a libtool from cctools?

From [libtool.c](https://opensource.apple.com/source/cctools/cctools-927.0.2/misc/libtool.c.auto.html):
> all of these values are initialized to reasonable defaults for deterministic
archives: archives that are consistent regardless of user, time, or umask
differences. Ordinarily, these values and will be set to more specific
ones when building new archives, but that can be suppressed using the
'-D' option or the ZERO_AR_DATE environment variable.

File: contrib/guix/libexec/build.sh [hunk 2]
```diff
@@ -63,6 +101,8 @@ for p in "${PATHS[@]}"; do
     fi
 done
 
+echo hee
```
Comment (fanquake) [inline contrib/guix/libexec/build.sh]: 🗣

**Review by fanquake — COMMENTED** [review]
Obviously concept ACK on macOS GUIX builds. Not going to comment on the interim system Clang commits here.

**Comment by dongcarl** [comment]
Update: this is still being actively worked on. Current hurdle: some weird clang internal search path logic.

**Comment by Sjors** [comment]
Tested 40ad76806b9f750d9c94675b3a136132bf1e6449. I (re)installed guix on an Ubuntu 20 machine and used `--no-substitutes` and `--bootstrap` to exercise my low time preference.

This did not go without hickups, e.g.:
* `guix install glibc-locales` (as root, this would fail without substitutes) and configuring `GUIX_LOCPATH` (as guix user) is not particularly well documented upstream
* during `guix-build.sh` the step `/gnu/store/36fgj9n3c8bmix2pd12kfaszi7bd5y7a-ghostscript-9.27.drv' failed` without too much actionable info. I tried dropping `--bootstrap`, but got the same error, so I ended up dropping `--no-substitutes` as well
* noticed a warning: `Computing Guix derivation for 'x86_64-linux'...  WARNING: (guix build emacs-build-system): imported module (guix build utils) overrides core binding `delete'`

Not sure what file is most useful to compare, but here's a few:

```
0aa66de467110bf45e7606d2696ea4417d4a94fdc59495646428c1ca05e2e744  distsrc-x86_64-linux-gnu/src/bitcoind
2b75f3955337b0345c7b6e33ad83185bba06ea96f1aedc30b7894c20af4489ee  distsrc-x86_64-apple-darwin16/Bitcoin-Qt.app/Contents/MacOS/Bitcoin-Qt
```

I can also still build and use depends on native macOS Big Sur. CI seems to struggle, maybe a spurious failure.

You may want to link to the macOS SDK extraction instructions (and that it needs to be untarred into depends/SDK's).

You may want to rebase (#20470) now that #20422 landed, since it changed `make deploy` (I just tried a rebase and it seems to behave at least on macOS itself).

Does this PR mean you can switch to the upstream GUIX repo? Or are there still patches? See https://github.com/bitcoin/bitcoin/pull/17595/commits/93439a71eda49fb69f1e82966a23a946733aa6fa

**Comment by dongcarl** [comment]
Many thanks for testing @Sjors! Your review is super helpful as I want to make this process as user-friendly as possible. Even for the folks building without substitutes! Cheers.

To start off, it seems like reproducibility is working!
```
$ sha256sum distsrc-x86_64-linux-gnu/src/bitcoind distsrc-x86_64-apple-darwin16/Bitcoin-Qt.app/Contents/MacOS/Bitcoin-Qt
0aa66de467110bf45e7606d2696ea4417d4a94fdc59495646428c1ca05e2e744  distsrc-x86_64-linux-gnu/src/bitcoind
2b75f3955337b0345c7b6e33ad83185bba06ea96f1aedc30b7894c20af4489ee  distsrc-x86_64-apple-darwin16/Bitcoin-Qt.app/Contents/MacOS/Bitcoin-Qt
```

Now onto your concerns:

> * `guix install glibc-locales` (as root, this would fail without substitutes) and configuring `GUIX_LOCPATH` (as guix user) is not particularly well documented upstream

Very true, I've had experience with this myself, I will write up some documentation. As for the glibc-locales failure, could you send me your `guix describe`?

> * during `guix-build.sh` the step `/gnu/store/36fgj9n3c8bmix2pd12kfaszi7bd5y7a-ghostscript-9.27.drv' failed` without too much actionable info. I tried dropping `--bootstrap`, but got the same error, so I ended up dropping `--no-substitutes` as well

I've encountered this a while ago, so this must be an outstanding bug. I will look into it! I also think it may be good for me to write down in the README how to report Guix build failures (Guix always saves a bzip'd full log file in `/var/log/guix` for failed builds).

> * noticed a warning: `Computing Guix derivation for 'x86_64-linux'...  WARNING: (guix build emacs-build-system): imported module (guix build utils) overrides core binding `delete'`

That warning is an old bug that has since been fixed here: https://issues.guix.gnu.org/41595. I will see how we can mitigate against this as well, may just have to bump my time-machine commit!

> You may want to link to the macOS SDK extraction instructions (and that it needs to be untarred into depends/SDK's).

Sure! I added a check for it in the script but I agree that it should be in the README as well!

> You may want to rebase (#20470) now that #20422 landed, since it changed make deploy (I just tried a rebase and it seems to behave at least on macOS itself).

Ah I did rebase that one, but not the cherry-picked commits in here. Will do!

> Does this PR mean you can switch to the upstream GUIX repo? Or are there still patches? See 93439a7

I want to switch over to upstream GUIX once everything is finalized. There could still be things that crop up I need to patch, but it's looking increasingly unlikely.

**Comment by Sjors** [comment]
guix describe (as root)
```
/gnu/store/pwcp239kjf7lnj5i4lkdzcfcxwcfyk72-bash-minimal-5.0.16/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
guix describe: error: failed to determine origin
hint: Perhaps this `guix' command was not obtained with `guix pull'? Its version string is 1.2.0.
```

(my Ubuntu install in general is a cluster f*** that I very much feel like reinstalling from scratch, FWIW)

> may just have to bump my time-machine commit

Happy to try again after that.

**Comment by dongcarl** [comment]
@Sjors Looking into a few things... Could you send me the file that's returned by:
```
$ guix build --log-file /gnu/store/36fgj9n3c8bmix2pd12kfaszi7bd5y7a-ghostscript-9.27.drv
```

Much appreciated.

**Comment by Sjors** [comment]
As user `guix` as well as `root`:
```
/gnu/store/pwcp239kjf7lnj5i4lkdzcfcxwcfyk72-bash-minimal-5.0.16/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
guix build: error: no build log for '/gnu/store/3i5qm6z0qihhjbdgs3pwj42gy63sf74h-ghostscript-9.27'
```

I ran the garbage collector `guix gc` and `git clean -xdff --exclude='/depends/SDKs/*'` and did another build (same hashes). 

By the way, before I ran that clean command I got:

```
DISTSRC directory '/bitcoin/distsrc-x86_64-linux-gnu' exists, probably because of previous builds... Aborting...
```

It would be nice if e.g. `guix-build.sh` checked this at the start, and maybe offers to nuke it.

**Comment by dongcarl** [comment]
@Sjors Would you mind giving me more info on your setup? I'm particularly interested in what distro and kernel version you were using when you encountered:

> guix install glibc-locales (as root, this would fail without substitutes)

**Comment by Sjors** [comment]
Ubuntu 20.04.1 LTS (GNU/Linux 5.4.0-33-generic x86_64)

I installed Guix as root using their `guix-install.sh` script. I have a user `guix` where I cloned this branch and from which I run the builds.

**Review by fanquake — COMMENTED** [review]
Built at 0e5c7bf5991a3e1c5e3df779a445e1690f067f5b. Tested the `.dmg` and Bitcoin-Qt. Is there a reason this is still draft?

```bash
d01c95e0040f55e610f2c99e81c307b4cd97bd2c9a9d14415709f8dcaaa5c380  output/bitcoin-0e5c7bf5991a-osx-unsigned.dmg
2f8b5077339a0eac72f78cae37437027608e0c88b3e250a4c82e73dc099dc224  output/bitcoin-0e5c7bf5991a-osx-unsigned.tar.gz
82b380107dbcde96ff66791c7b512498ecefc5bc2f2428390cccd83b8fce2bb4  output/bitcoin-0e5c7bf5991a-x86_64-apple-darwin18.tar.gz
107a81d69796b148360594c5c07a646155a1f3b14051a7b804fe27831f0f833c  output/src/bitcoin-0e5c7bf5991a.tar.gz
```

File: contrib/guix/README.md [hunk 3]
```diff
@@ -105,7 +105,8 @@ find output/ -type f -print0 | sort -z | xargs -r0 sha256sum
 
   Override the space-separated list of platform triples for which to perform a
   bootstrappable build. _(defaults to "x86\_64-linux-gnu arm-linux-gnueabihf
-  aarch64-linux-gnu riscv64-linux-gnu x86_64-w64-mingw32")_
+  aarch64-linux-gnu riscv64-linux-gnu x86_64-w64-mingw32
+  x86_64-apple-darwin16")_
```
[the remark this answers is not in the document]
Comment (dongcarl) [reply contrib/guix/README.md]: Thanks! Fixed as of 05a31a95832b592ee565d3a429afcf6cbf5ae672

File: contrib/guix/libexec/build.sh [hunk 4]
```diff
@@ -199,17 +245,22 @@ mkdir -p "$DISTSRC"
...
 
     sed -i.old 's/-lstdc++ //g' config.status libtool src/univalue/config.status src/univalue/libtool
 
     # Build Bitcoin Core
     make --jobs="$MAX_JOBS" ${V:+V=1}
 
-    # Perform basic ELF security checks on a series of executables.
-    make -C src --jobs=1 check-security ${V:+V=1}
+    case "$HOST" in
+        *darwin*) ;;
```
[the remark this answers is not in the document]
Comment (dongcarl) [reply contrib/guix/libexec/build.sh]: Originally I skipped them because we don't perform them in gitian osx builds, which I thought had some significance. As of 05a31a95832b592ee565d3a429afcf6cbf5ae672 they are enabled for all builds in Guix. Sweet!

**Comment by dongcarl** [comment]
Pushed 0e5c7bf5991a3e1c5e3df779a445e1690f067f5b -> 05a31a95832b592ee565d3a429afcf6cbf5ae672
- Addressed https://github.com/bitcoin/bitcoin/pull/17920#discussion_r556184995 and https://github.com/bitcoin/bitcoin/pull/17920#discussion_r556185284

**Comment by fanquake** [comment]
05a31a95832b592ee565d3a429afcf6cbf5ae672
```bash
find output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
3f2762da34e6681471a7ede52ff7739e0e5ad64dd4584ae1e6d4af47456c2a98  output/bitcoin-05a31a95832b-osx-unsigned.dmg
ffc15b01f5028bf15fda61c1b3d785c2ba6e72fdfdd3292ea7f7d7f0b5230f9f  output/bitcoin-05a31a95832b-osx-unsigned.tar.gz
538ba85976b7dec9353d9e246346cac792916ed2a59360c5ded8c7e4ddebe8b7  output/bitcoin-05a31a95832b-osx64.tar.gz
21118fd0183fc430fe6de82bbcd5ab353643fc51a280a11cb188e9e4ae7c7cd5  output/src/bitcoin-05a31a95832b.tar.gz
```

**Comment by dongcarl** [comment]
I'm getting matching hashes!

```
find output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
3f2762da34e6681471a7ede52ff7739e0e5ad64dd4584ae1e6d4af47456c2a98  output/bitcoin-05a31a95832b-osx-unsigned.dmg
ffc15b01f5028bf15fda61c1b3d785c2ba6e72fdfdd3292ea7f7d7f0b5230f9f  output/bitcoin-05a31a95832b-osx-unsigned.tar.gz
538ba85976b7dec9353d9e246346cac792916ed2a59360c5ded8c7e4ddebe8b7  output/bitcoin-05a31a95832b-osx64.tar.gz
21118fd0183fc430fe6de82bbcd5ab353643fc51a280a11cb188e9e4ae7c7cd5  output/src/bitcoin-05a31a95832b.tar.gz
```

**Comment by dongcarl** [comment]
Pushed 05a31a95832b592ee565d3a429afcf6cbf5ae672 -> f1694757ddbcb3635213b085e864851e285c8c12
- Rebased over master
- Fixed typo

Hashes:
```
2d17b6c787f91bf32e5d3732a6a054df1940d26030fc19d845ca0b472283a1bb  output/bitcoin-f1694757ddbc-osx-unsigned.dmg
c5a8207f4d76928b88ac9dfed61197be9bfd62ba93821898a9c90e1eb8a5a7d5  output/bitcoin-f1694757ddbc-osx-unsigned.tar.gz
bb972ac7f0f03c31d30c15cc9667530cd25f2237f4eb9d32cd3a1ea17ce1ff6a  output/bitcoin-f1694757ddbc-osx64.tar.gz
262cb14aabffdb98260ca53e5d420b14f4242c83e414963abf9c4798eba899b6  output/src/bitcoin-f1694757ddbc.tar.gz
```

**Comment by fanquake** [comment]
ACK f1694757ddbcb3635213b085e864851e285c8c12 - I think we can make some small usability improvements, but this is ok to merge now.

```bash
find output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
2d17b6c787f91bf32e5d3732a6a054df1940d26030fc19d845ca0b472283a1bb  output/bitcoin-f1694757ddbc-osx-unsigned.dmg
c5a8207f4d76928b88ac9dfed61197be9bfd62ba93821898a9c90e1eb8a5a7d5  output/bitcoin-f1694757ddbc-osx-unsigned.tar.gz
bb972ac7f0f03c31d30c15cc9667530cd25f2237f4eb9d32cd3a1ea17ce1ff6a  output/bitcoin-f1694757ddbc-osx64.tar.gz
262cb14aabffdb98260ca53e5d420b14f4242c83e414963abf9c4798eba899b6  output/src/bitcoin-f1694757ddbc.tar.gz
```

File: contrib/guix/manifest.scm [hunk 5]
```diff
@@ -195,8 +222,8 @@ chain for " target " development."))
         python-3.7
         ;; Git
         git
-        ;; Native gcc 9 toolchain targeting glibc 2.27
-        (make-gcc-toolchain gcc-9 glibc-2.27))
+        ;; Native gcc 7 toolchain
+        gcc-toolchain-7)
```
Comment (MarcoFalke) [inline contrib/guix/manifest.scm:226]: Why is this changed, seems unrelated to macos, no? Does this break gcc cross builds for our linux archs that need gcc 8 at least?

**Comment by Sjors** [comment]
Post merge test: I get the same checksums!

File: contrib/guix/manifest.scm (hunk 5 above)
Comment (MarcoFalke) [reply contrib/guix/manifest.scm:226]: Btw, this is the comment I was referring to: https://github.com/bitcoin/bitcoin/pull/13665#issuecomment-405601228

I presume that guix is using a gcc-7 that has a fix backported?
